runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -6579,6 +6579,7 @@ TR::Register *J9::Power::TreeEvaluator::VMnewEvaluator(TR::Node *node, TR::CodeG
          TR_ExtraInfoForNew *initInfo = node->getSymbolReference()->getExtraInfo();
 
          static bool disableFastArrayZeroInit = (feGetEnv("TR_DisableFastArrayZeroInit") != NULL);
+         iCursor = generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, zeroReg, 0, iCursor);
 
          if (!node->canSkipZeroInitialization() && (initInfo == NULL || initInfo->numZeroInitSlots > 0))
             {
```

Comment:
Is this instruction needed? Line 6467 has the comment

```// On return, zeroReg is set to 0 if needZeroInit is true```

prior to the `genHeapAlloc()` call and that call looks like it is always executed. Here you're inside a block guarded by `if (needZeroInit && (!isConstantZeroLenArrayAlloc))` so it looks like you will always have a properly cleared `zeroReg` at this point.

You should see two `li rX, 0` instructions in your trace logs currently if the comment is correct, in which case we don't need the second one.

Reply:
@ymanton Currently in genHeapAlloc, zeroReg gets initialized in different scenarios.
1. With realTimeGC() option is set in LineNo 5352
2. With !dualTLH mode LineNo 5531
3. With !usingTLH mode LineNo 5864

I have this from the instructions generated:
```
0x7830c6f7ea78 {java/util/Hashtable.<init>} +31                  |||  d81570e8 ld        r3, 0x15d8(r16) JITTOC 0x14da00 Ptr                                                            
0x7830c6f7ea7c {java/util/Hashtable.<init>} +32                  |||  60000fe9 ld        r8, 0x60(r15) J9VMThread.heapAlloc // TLH Alloc                                                
0x7830c6f7ea80 {java/util/Hashtable.<init>} +33                  |||  68008fe8 ld        r4, 0x68(r15) J9VMThread.heapTop                                                               
0x7830c6f7ea84 {java/util/Hashtable.<init>} +34                  |||  40000838 addi      r0, r8, 0x40                                                                                   
0x7830c6f7ea88 {java/util/Hashtable.<init>} +35                  |||  4020207c cmpld     r0, r4                                                                                         0x7830c6f7ea8c {java/util/Hashtable.<init>} +36                  |||  a4018141 bgt       0x7830c6f7ec30 C>> +141                                                                        
0x7830c6f7ea90 {java/util/Hashtable.<init>} +37                  |||  60000ff8 std       r0, 0x60(r15) J9VMThread.heapAlloc
0x7830c6f7ea94 {java/util/Hashtable.<init>} +38                  |||  e01590e8 ld        r4, 0x15e0(r16) JITTOC 0x14df00 Ptr                                     
0x7830c6f7ea98 {java/util/Hashtable.<init>} +39                  |||  00008890 stw       r4, 0(r8)                                                                                      0x7830c6f7ea9c {java/util/Hashtable.<init>} +40                  |||  04004890 stw       r2, 4(r8)                                                                                      
0x7830c6f7eaa0 {java/util/Hashtable.<init>} +41                  |||  10006839 addi      r11, r8, 0x10           
0x7830c6f7eaa4 {java/util/Hashtable.<init>} +42                  |||  080068f9 std       r11, 8(r8)                                                                                     
0x7830c6f7eaa8 {java/util/Hashtable.<init>} +43                  |||  0000e038 li        r7, 0                               
0x7830c6f7eaac {java/util/Hashtable.<init>} +44                  |||  1000e8f8 std       r7, 0x10(r8)                                    
0x7830c6f7eab0 {java/util/Hashtable.<init>} +45                  |||  1800e8f8 std       r7, 0x18(r8)                    
0x7830c6f7eab4 {java/util/Hashtable.<init>} +46                  |||  2000e8f8 std       r7, 0x20(r8)
0x7830c6f7eab8 {java/util/Hashtable.<init>} +47                  |||  2800e8f8 std       r7, 0x28(r8)            
0x7830c6f7eabc {java/util/Hashtable.<init>} +48                  |||  3000e8f8 std       r7, 0x30(r8)              
0x7830c6f7eac0 {java/util/Hashtable.<init>} +49                  |||  3800e8f8 std       r7, 0x38(r8)
```